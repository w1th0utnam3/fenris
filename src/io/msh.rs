//! Support for loading Gmsh MSH 4.1 files as [`Mesh`]es.
//!
//! This module supports Gmsh MSH files with the file format version 4.1 as specified in the
//! [Gmsh documentation](https://gmsh.info/doc/texinfo/gmsh.html#MSH-file-format). All element types
//! that implement the [`MshConnectivity`] trait are supported.
//!
//! Example usage:
//! ```
//! use nalgebra::U2;
//! use fenris::connectivity::Tri3d2Connectivity;
//! use fenris::io::msh::load_msh_from_file;
//!
//! let mesh = load_msh_from_file::<f64, U2, Tri3d2Connectivity, _>(
//!     "assets/meshes/square_tri3_4.msh").unwrap();
//!
//! assert_eq!(mesh.vertices().len(), 5);
//! assert_eq!(mesh.connectivity().len(), 4);
//! ```

use crate::connectivity::{
    Hex27Connectivity, Hex8Connectivity, Quad4d2Connectivity, Quad9d2Connectivity, Tet10Connectivity, Tet4Connectivity,
    Tri3d2Connectivity, Tri3d3Connectivity, Tri6d2Connectivity,
};
use crate::mesh::Mesh;
use eyre::{eyre, Context};
use nalgebra::allocator::Allocator;
use nalgebra::{DefaultAllocator, DimName, OPoint, RealField};
use num::ToPrimitive;
use std::path::Path;

/// Loads a [`Mesh`] from a Gmsh MSH file at the given path.
pub fn load_msh_from_file<T, D, C, P: AsRef<Path>>(file_path: P) -> eyre::Result<Mesh<T, D, C>>
where
    T: RealField,
    D: DimName,
    C: MshConnectivity,
    DefaultAllocator: Allocator<T, D>,
{
    let msh_bytes = std::fs::read(file_path).wrap_err("failed to read file")?;
    load_msh_from_bytes(&msh_bytes).wrap_err("failed to load mesh from msh file")
}

/// Loads a [`Mesh`] by parsing the given bytes as a Gmsh MSH file.
pub fn load_msh_from_bytes<T, D, C>(bytes: &[u8]) -> eyre::Result<Mesh<T, D, C>>
where
    T: RealField,
    D: DimName,
    C: MshConnectivity,
    DefaultAllocator: Allocator<T, D>,
{
    let mut msh_file = mshio::parse_msh_bytes(bytes).map_err(|e| eyre!("failed to parse msh file: {}", e))?;

    let msh_nodes = msh_file
        .data
        .nodes
        .take()
        .ok_or(eyre!("MSH file does not contain nodes"))?;
    let msh_elements = msh_file
        .data
        .elements
        .take()
        .ok_or(eyre!("MSH file does not contain elements"))?;

    let mut vertices = Vec::new();
    let mut connectivity = Vec::new();

    // Ensure that at least one element block matches the target mesh connectivity
    if !msh_elements
        .element_blocks
        .iter()
        .any(|block| element_block_matches_connectivity::<C, _>(block))
    {
        return Err(eyre!(
            "MSH file does not contain an element block of the requested type (type {:?} of with reference/entity dim {})",
            C::msh_element_type(),
            C::reference_dim()
        ));
    }

    // Collect all mesh vertices
    for node_block in &msh_nodes.node_blocks {
        let block_vertices = vertices_from_node_block(node_block)?;
        vertices.extend(block_vertices);
    }

    if vertices.len()
        != msh_nodes
            .num_nodes
            .to_usize()
            .expect("failed to convert num_nodes to usize")
    {
        return Err(eyre!(
            "only {} vertices were read but msh file claims to contain {} nodes",
            vertices.len(),
            msh_nodes.num_nodes
        ));
    }

    // Collect all connectivity matching the target connectivity
    for element_block in &msh_elements.element_blocks {
        let block_connectivity = connectivity_from_element_block(element_block)?;
        connectivity.extend(block_connectivity);
    }

    Ok(Mesh::from_vertices_and_connectivity(vertices, connectivity))
}

/// Tries to convert a `mshio::NodeBlock` to a `Vec<OPoint<T, D>>`.
fn vertices_from_node_block<T, D, F, I>(node_block: &mshio::NodeBlock<u64, I, F>) -> eyre::Result<Vec<OPoint<T, D>>>
where
    T: RealField,
    D: DimName,
    F: mshio::MshFloatT,
    I: mshio::MshIntT,
    DefaultAllocator: Allocator<T, D>,
{
    // Ensure that node tags are consecutive
    if node_block.node_tags.is_some() {
        return Err(eyre!(
            "node block tags are not consecutive in msh file (sparse tags are not supported)"
        ));
    }

    // Note: The following check is commented out because it can lead to missing nodes.
    //  Probable explanation:
    //  The MSH `node_block`'s `entity_dim` does not seem to correspond to the geometrical
    //  dimension of the points. Rather it seems to correspond to the dimension of the "physical"
    //  object represented by the node block.
    //  When creating primitives in Gmsh for example, the nodes of a triangulation of a sphere are
    //  divided into node_blocks representing its equator, surface and volume and all of them are
    //  referenced by the volumetric elements.
    //  In addition, all node blocks have to be read anyway in order for the global `node_tag` indexing to
    //  work consistently.
    // TODO: Revisit this after some more experience with loading MSH files and delete this.
    /*
    if node_block
        .entity_dim
        .to_usize()
        .ok_or_else(|| eyre!("error converting node block entity dimension to usize"))?
        != D::dim()
    {
        return Err(eyre!("node block entity does not have the right dimension for this mesh"));
    }
    */

    let mut vertices = Vec::with_capacity(node_block.nodes.len());

    // Convert MSH vertices to points
    for node in &node_block.nodes {
        vertices.push(point_from_msh_node(node)?);
    }

    Ok(vertices)
}

/// Tries to convert a `mshio::ElementBlock` to a `Vec<Connectivity>`.
fn connectivity_from_element_block<C, I>(element_block: &mshio::ElementBlock<u64, I>) -> eyre::Result<Vec<C>>
where
    C: MshConnectivity,
    I: mshio::MshIntT,
{
    // Ensure that element tags are consecutive
    if element_block.element_tags.is_some() {
        return Err(eyre!(
            "element block tags are not consecutive in msh file (sparse tags are not supported)"
        ));
    }

    if !element_block_matches_connectivity::<C, _>(element_block) {
        // Just ignore blocks that don't match the requested connectivity
        return Ok(Vec::new());
    } else {
        let mut connectivity = Vec::with_capacity(element_block.elements.len());
        let requested_nodes = C::msh_element_type()
            .nodes()
            .map_err(|_| eyre!("unimplemented element type requested"))?;

        for element in &element_block.elements {
            if element.nodes.len() < requested_nodes {
                return Err(eyre!("not enough nodes to initialize connectivity"));
            }
            connectivity.push(C::try_connectivity_from_msh_element(element)?);
        }

        return Ok(connectivity);
    }
}

/// Returns whether the given element block contains elements corresponding to the specified connectivity.
fn element_block_matches_connectivity<C, I>(element_block: &mshio::ElementBlock<u64, I>) -> bool
where
    C: MshConnectivity,
    I: mshio::MshIntT,
{
    element_block.element_type == C::msh_element_type()
        && element_block
            .entity_dim
            .to_usize()
            .expect("failed to convert element block dimension to usize")
            == C::reference_dim()
}

macro_rules! f_to_t {
    ($component:expr) => {
        T::from_f64(
            $component
                .to_f64()
                .ok_or_else(|| eyre!("failed to convert coordinate to f64"))?,
        )
        .ok_or_else(|| eyre!("failed to convert node coordinate from f64 to target mesh real type"))?
    };
}

fn point_from_msh_node<T, D, F>(node: &mshio::Node<F>) -> eyre::Result<OPoint<T, D>>
where
    T: RealField,
    D: DimName,
    F: mshio::MshFloatT,
    DefaultAllocator: Allocator<T, D>,
{
    // TODO: Ensure that components i < D are zero?
    let mut point = OPoint::origin();
    point[0] = f_to_t!(node.x);
    if D::dim() > 1 {
        point[1] = f_to_t!(node.y);
    }
    if D::dim() > 2 {
        point[2] = f_to_t!(node.z);
    }

    Ok(point)
}

/// Allows conversion from `mshio::Element`s to connectivity types used in `fenris`.
pub trait MshConnectivity
where
    Self: Sized,
{
    /// Returns the MSH element type corresponding to this connectivity.
    fn msh_element_type() -> mshio::ElementType;
    /// Returns the reference dimension of this connectivity (corresponds to MSH entity dimension).
    fn reference_dim() -> usize;
    /// Tries to construct the element connectivity from the given MSH element.
    fn try_connectivity_from_msh_element(element: &mshio::Element<u64>) -> eyre::Result<Self>;
}

macro_rules! impl_msh_connectivity {
    ($connectivity:ident, $msh_type:ident, num_nodes = $num_nodes:literal) => {
        impl MshConnectivity for $connectivity {
            fn msh_element_type() -> mshio::ElementType {
                mshio::ElementType::$msh_type
            }

            fn reference_dim() -> usize {
                use crate::element::ElementConnectivity;
                <$connectivity as ElementConnectivity<f64>>::ReferenceDim::dim()
            }

            fn try_connectivity_from_msh_element(element: &mshio::Element<u64>) -> eyre::Result<Self> {
                assert_eq!(
                    element.nodes.len(),
                    $num_nodes,
                    "number of msh element nodes have to match with connectivity"
                );
                let mut nodes = [0; $num_nodes];
                for i in 0..$num_nodes {
                    nodes[i] = element.nodes[i] as usize - 1;
                }
                Ok(Self(nodes))
            }
        }
    };
}

impl_msh_connectivity!(Tri3d2Connectivity, Tri3, num_nodes = 3);
impl_msh_connectivity!(Tri3d3Connectivity, Tri3, num_nodes = 3);
impl_msh_connectivity!(Tri6d2Connectivity, Tri6, num_nodes = 6);
impl_msh_connectivity!(Quad4d2Connectivity, Qua4, num_nodes = 4);
impl_msh_connectivity!(Quad9d2Connectivity, Qua9, num_nodes = 9);
impl_msh_connectivity!(Tet4Connectivity, Tet4, num_nodes = 4);
impl_msh_connectivity!(Tet10Connectivity, Tet10, num_nodes = 10);
impl_msh_connectivity!(Hex8Connectivity, Hex8, num_nodes = 8);
impl_msh_connectivity!(Hex27Connectivity, Hex27, num_nodes = 27);

// The following connectivities do not implement ElementConnectivity yet
//impl_msh_connectivity!(Tri6d3Connectivity, Tri6, num_nodes = 6);
//impl_msh_connectivity!(Quad4d3Connectivity, Qua4, num_nodes = 4);
//impl_msh_connectivity!(Quad9d3Connectivity, Qua9, num_nodes = 9);
