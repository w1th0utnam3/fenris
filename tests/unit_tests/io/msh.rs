use crate::export_mesh_vtk;
use fenris::connectivity::{
    Hex27Connectivity, Hex8Connectivity, Quad4d2Connectivity, Quad9d2Connectivity, Tet10Connectivity, Tet4Connectivity,
    Tri3d2Connectivity, Tri3d3Connectivity, Tri6d2Connectivity,
};
use fenris::io::msh::load_msh_from_file;
use insta::assert_debug_snapshot;
use nalgebra::{U2, U3};

#[test]
fn load_msh_sphere_tet4_large() -> eyre::Result<()> {
    let mesh = load_msh_from_file::<f64, U3, Tet4Connectivity, _>("assets/meshes/sphere_tet4_593.msh")?;

    assert_eq!(mesh.vertices().len(), 183);
    assert_eq!(mesh.connectivity().len(), 593);

    export_mesh_vtk("io_msh", "load_msh_sphere_tet4_large", &mesh);
    assert_debug_snapshot!(mesh);

    Ok(())
}

#[test]
fn load_msh_rect_tri3d2_large() -> eyre::Result<()> {
    let mesh = load_msh_from_file::<f64, U2, Tri3d2Connectivity, _>("assets/meshes/rectangle_tri3_110.msh")?;

    assert_eq!(mesh.vertices().len(), 70);
    assert_eq!(mesh.connectivity().len(), 110);

    export_mesh_vtk("io_msh", "load_msh_rect_tri3d2_large", &mesh);
    assert_debug_snapshot!(mesh);

    Ok(())
}

#[test]
fn load_msh_rect_tri3d3_large() -> eyre::Result<()> {
    // Loading a 2D triangle mesh to a 3D triangle mesh should work
    let mesh = load_msh_from_file::<f64, U3, Tri3d3Connectivity, _>("assets/meshes/rectangle_tri3_110.msh")?;

    assert_eq!(mesh.vertices().len(), 70);
    assert_eq!(mesh.connectivity().len(), 110);

    export_mesh_vtk("io_msh", "load_msh_rect_tri3d3_large", &mesh);
    assert_debug_snapshot!(mesh);

    Ok(())
}

#[test]
fn load_msh_square_quad4d2_large() -> eyre::Result<()> {
    let mesh = load_msh_from_file::<f64, U2, Quad4d2Connectivity, _>("assets/meshes/square_quad4_79.msh")?;

    assert_eq!(mesh.vertices().len(), 96);
    assert_eq!(mesh.connectivity().len(), 79);

    export_mesh_vtk("io_msh", "load_msh_square_quad4d2_large", &mesh);
    assert_debug_snapshot!(mesh);

    Ok(())
}

#[test]
fn load_msh_square_tri3d2() -> eyre::Result<()> {
    let mesh = load_msh_from_file::<f64, U2, Tri3d2Connectivity, _>("assets/meshes/square_tri3_4.msh")?;

    assert_eq!(mesh.vertices().len(), 5);
    assert_eq!(mesh.connectivity().len(), 4);

    export_mesh_vtk("io_msh", "load_msh_square_tri3d2", &mesh);
    assert_debug_snapshot!(mesh);

    Ok(())
}

#[test]
fn load_msh_square_tri6d2() -> eyre::Result<()> {
    let mesh = load_msh_from_file::<f64, U2, Tri6d2Connectivity, _>("assets/meshes/square_tri6_4.msh")?;

    assert_eq!(mesh.vertices().len(), 13);
    assert_eq!(mesh.connectivity().len(), 4);

    export_mesh_vtk("io_msh", "load_msh_square_tri6d2", &mesh);
    assert_debug_snapshot!(mesh);

    Ok(())
}

#[test]
fn load_msh_square_quad4d2() -> eyre::Result<()> {
    let mesh = load_msh_from_file::<f64, U2, Quad4d2Connectivity, _>("assets/meshes/square_quad4_4.msh")?;

    assert_eq!(mesh.vertices().len(), 9);
    assert_eq!(mesh.connectivity().len(), 4);

    export_mesh_vtk("io_msh", "load_msh_square_quad4d2", &mesh);
    assert_debug_snapshot!(mesh);

    Ok(())
}

#[test]
fn load_msh_square_quad9d2() -> eyre::Result<()> {
    let mesh = load_msh_from_file::<f64, U2, Quad9d2Connectivity, _>("assets/meshes/square_quad9_4.msh")?;

    assert_eq!(mesh.vertices().len(), 25);
    assert_eq!(mesh.connectivity().len(), 4);

    export_mesh_vtk("io_msh", "load_msh_square_quad9d2", &mesh);
    assert_debug_snapshot!(mesh);

    Ok(())
}

#[test]
fn load_msh_cube_tet4() -> eyre::Result<()> {
    let mesh = load_msh_from_file::<f64, U3, Tet4Connectivity, _>("assets/meshes/cube_tet4_24.msh")?;

    assert_eq!(mesh.vertices().len(), 14);
    assert_eq!(mesh.connectivity().len(), 24);

    export_mesh_vtk("io_msh", "load_msh_cube_tet4", &mesh);
    assert_debug_snapshot!(mesh);

    Ok(())
}

#[test]
fn load_msh_cube_tet10() -> eyre::Result<()> {
    let mesh = load_msh_from_file::<f64, U3, Tet10Connectivity, _>("assets/meshes/cube_tet10_24.msh")?;

    assert_eq!(mesh.vertices().len(), 63);
    assert_eq!(mesh.connectivity().len(), 24);

    export_mesh_vtk("io_msh", "load_msh_cube_tet10", &mesh);
    assert_debug_snapshot!(mesh);

    Ok(())
}

#[test]
fn load_msh_cube_hex8() -> eyre::Result<()> {
    let mesh = load_msh_from_file::<f64, U3, Hex8Connectivity, _>("assets/meshes/cube_hex8_8.msh")?;

    assert_eq!(mesh.vertices().len(), 27);
    assert_eq!(mesh.connectivity().len(), 8);

    export_mesh_vtk("io_msh", "load_msh_cube_hex8", &mesh);
    assert_debug_snapshot!(mesh);

    Ok(())
}

#[test]
fn load_msh_cube_hex27() -> eyre::Result<()> {
    let mesh = load_msh_from_file::<f64, U3, Hex27Connectivity, _>("assets/meshes/cube_hex27_8.msh")?;

    assert_eq!(mesh.vertices().len(), 125);
    assert_eq!(mesh.connectivity().len(), 8);

    export_mesh_vtk("io_msh", "load_msh_cube_hex27", &mesh);
    assert_debug_snapshot!(mesh);

    Ok(())
}
