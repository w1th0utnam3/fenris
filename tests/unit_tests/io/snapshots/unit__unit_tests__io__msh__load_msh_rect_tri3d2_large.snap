---
source: tests/unit_tests/io/msh.rs
expression: mesh
---
Mesh {
    vertices: [
        [
            -0.3,
            -0.3,
        ],
        [
            0.7,
            -0.3,
        ],
        [
            0.7,
            0.2,
        ],
        [
            -0.3,
            0.2,
        ],
        [
            -0.1888888888888891,
            -0.3,
        ],
        [
            -0.07777777777777825,
            -0.3,
        ],
        [
            0.03333333333333294,
            -0.3,
        ],
        [
            0.1444444444444442,
            -0.3,
        ],
        [
            0.2555555555555554,
            -0.3,
        ],
        [
            0.3666666666666664,
            -0.3,
        ],
        [
            0.4777777777777777,
            -0.3,
        ],
        [
            0.5888888888888888,
            -0.3,
        ],
        [
            0.7,
            -0.2000000000000002,
        ],
        [
            0.7,
            -0.1000000000000001,
        ],
        [
            0.7,
            -1.110223024625157e-16,
        ],
        [
            0.7,
            0.09999999999999998,
        ],
        [
            0.588888888888889,
            0.2,
        ],
        [
            0.4777777777777782,
            0.2,
        ],
        [
            0.366666666666667,
            0.2,
        ],
        [
            0.2555555555555558,
            0.2,
        ],
        [
            0.1444444444444446,
            0.2,
        ],
        [
            0.03333333333333355,
            0.2,
        ],
        [
            -0.07777777777777772,
            0.2,
        ],
        [
            -0.1888888888888889,
            0.2,
        ],
        [
            -0.3,
            0.1000000000000002,
        ],
        [
            -0.3,
            1.665334536937735e-16,
        ],
        [
            -0.3,
            -0.09999999999999987,
        ],
        [
            -0.3,
            -0.2,
        ],
        [
            0.6181853054877119,
            -0.0415974196800525,
        ],
        [
            -0.2184200639486318,
            -0.05004133303915795,
        ],
        [
            0.2017105453434146,
            -0.2002352435348832,
        ],
        [
            0.3104891941360209,
            0.1029078944520922,
        ],
        [
            0.08834248811904713,
            0.1038514358620944,
        ],
        [
            0.4416365682405676,
            -0.2094183433526482,
        ],
        [
            -0.02487999700958349,
            -0.1993290750383707,
        ],
        [
            0.5363367134450407,
            0.09963349191309573,
        ],
        [
            -0.1376937270386464,
            0.09918316830157284,
        ],
        [
            -0.02320268433312986,
            0.1034937157688774,
        ],
        [
            0.03225151084425054,
            0.00824487064593734,
        ],
        [
            0.1430952130559973,
            0.007257297698020186,
        ],
        [
            0.08670007502345552,
            -0.08790564410859014,
        ],
        [
            -0.02366509376319328,
            -0.08669558283174966,
        ],
        [
            0.1971963049208412,
            -0.08916950506662782,
        ],
        [
            0.253922991933277,
            0.006317558455476907,
        ],
        [
            0.3081255793967473,
            -0.09077591339602184,
        ],
        [
            0.3646194671681309,
            0.005510616989843405,
        ],
        [
            0.4185164091280256,
            -0.0918398591757682,
        ],
        [
            0.4701913192082958,
            0.01124911446365311,
        ],
        [
            0.4222222222222225,
            0.1037749551350623,
        ],
        [
            0.3165578285233138,
            -0.2010429173546686,
        ],
        [
            0.51294097640968,
            -0.1461571739671005,
        ],
        [
            -0.07204502561708118,
            0.003996362274948255,
        ],
        [
            -0.1428703653664571,
            -0.2133259091633978,
        ],
        [
            0.0882666209765149,
            -0.2041342216817793,
        ],
        [
            0.1993083145407238,
            0.1033890310779473,
        ],
        [
            0.6141673003725829,
            -0.1500000000000002,
        ],
        [
            -0.2150687950126897,
            -0.1499999999999999,
        ],
        [
            -0.1217096898957324,
            -0.09916445375994822,
        ],
        [
            -0.2228047796821482,
            0.04659250577317239,
        ],
        [
            0.6177454205875057,
            0.04994026324231237,
        ],
        [
            0.5466426593556569,
            -0.2259292528866248,
        ],
        [
            0.5296132320729046,
            -0.06139610197017452,
        ],
        [
            0.628594204584287,
            0.1299147510310816,
        ],
        [
            -0.2286828846737035,
            0.1291590499816862,
        ],
        [
            -0.2256650626505246,
            -0.2291340202289778,
        ],
        [
            -0.1550184885524905,
            0.001244361478090254,
        ],
        [
            0.03088955984017527,
            -0.1445718351398361,
        ],
        [
            0.6299397697234258,
            -0.2351858505773251,
        ],
        [
            0.141299000533621,
            -0.1452945975736307,
        ],
        [
            0.5531073904611943,
            0.0116360736783649,
        ],
    ],
    connectivity: [
        Tri3d2Connectivity(
            [
                30,
                49,
                44,
            ],
        ),
        Tri3d2Connectivity(
            [
                30,
                44,
                42,
            ],
        ),
        Tri3d2Connectivity(
            [
                46,
                49,
                33,
            ],
        ),
        Tri3d2Connectivity(
            [
                44,
                49,
                46,
            ],
        ),
        Tri3d2Connectivity(
            [
                41,
                57,
                34,
            ],
        ),
        Tri3d2Connectivity(
            [
                34,
                57,
                52,
            ],
        ),
        Tri3d2Connectivity(
            [
                10,
                60,
                33,
            ],
        ),
        Tri3d2Connectivity(
            [
                33,
                60,
                50,
            ],
        ),
        Tri3d2Connectivity(
            [
                35,
                48,
                47,
            ],
        ),
        Tri3d2Connectivity(
            [
                51,
                57,
                41,
            ],
        ),
        Tri3d2Connectivity(
            [
                47,
                48,
                45,
            ],
        ),
        Tri3d2Connectivity(
            [
                45,
                48,
                31,
            ],
        ),
        Tri3d2Connectivity(
            [
                28,
                61,
                55,
            ],
        ),
        Tri3d2Connectivity(
            [
                33,
                50,
                46,
            ],
        ),
        Tri3d2Connectivity(
            [
                43,
                45,
                31,
            ],
        ),
        Tri3d2Connectivity(
            [
                33,
                49,
                9,
            ],
        ),
        Tri3d2Connectivity(
            [
                9,
                49,
                8,
            ],
        ),
        Tri3d2Connectivity(
            [
                17,
                48,
                35,
            ],
        ),
        Tri3d2Connectivity(
            [
                55,
                61,
                50,
            ],
        ),
        Tri3d2Connectivity(
            [
                34,
                52,
                5,
            ],
        ),
        Tri3d2Connectivity(
            [
                23,
                36,
                22,
            ],
        ),
        Tri3d2Connectivity(
            [
                17,
                35,
                16,
            ],
        ),
        Tri3d2Connectivity(
            [
                38,
                51,
                41,
            ],
        ),
        Tri3d2Connectivity(
            [
                37,
                38,
                32,
            ],
        ),
        Tri3d2Connectivity(
            [
                8,
                49,
                30,
            ],
        ),
        Tri3d2Connectivity(
            [
                31,
                54,
                43,
            ],
        ),
        Tri3d2Connectivity(
            [
                21,
                37,
                32,
            ],
        ),
        Tri3d2Connectivity(
            [
                36,
                37,
                22,
            ],
        ),
        Tri3d2Connectivity(
            [
                46,
                47,
                45,
            ],
        ),
        Tri3d2Connectivity(
            [
                44,
                46,
                45,
            ],
        ),
        Tri3d2Connectivity(
            [
                36,
                51,
                37,
            ],
        ),
        Tri3d2Connectivity(
            [
                10,
                33,
                9,
            ],
        ),
        Tri3d2Connectivity(
            [
                6,
                34,
                5,
            ],
        ),
        Tri3d2Connectivity(
            [
                5,
                52,
                4,
            ],
        ),
        Tri3d2Connectivity(
            [
                38,
                41,
                40,
            ],
        ),
        Tri3d2Connectivity(
            [
                37,
                51,
                38,
            ],
        ),
        Tri3d2Connectivity(
            [
                30,
                53,
                7,
            ],
        ),
        Tri3d2Connectivity(
            [
                18,
                48,
                17,
            ],
        ),
        Tri3d2Connectivity(
            [
                31,
                48,
                18,
            ],
        ),
        Tri3d2Connectivity(
            [
                6,
                53,
                34,
            ],
        ),
        Tri3d2Connectivity(
            [
                20,
                54,
                19,
            ],
        ),
        Tri3d2Connectivity(
            [
                44,
                45,
                43,
            ],
        ),
        Tri3d2Connectivity(
            [
                38,
                40,
                39,
            ],
        ),
        Tri3d2Connectivity(
            [
                43,
                54,
                39,
            ],
        ),
        Tri3d2Connectivity(
            [
                39,
                54,
                32,
            ],
        ),
        Tri3d2Connectivity(
            [
                32,
                54,
                20,
            ],
        ),
        Tri3d2Connectivity(
            [
                19,
                54,
                31,
            ],
        ),
        Tri3d2Connectivity(
            [
                22,
                37,
                21,
            ],
        ),
        Tri3d2Connectivity(
            [
                38,
                39,
                32,
            ],
        ),
        Tri3d2Connectivity(
            [
                7,
                53,
                6,
            ],
        ),
        Tri3d2Connectivity(
            [
                42,
                44,
                43,
            ],
        ),
        Tri3d2Connectivity(
            [
                8,
                30,
                7,
            ],
        ),
        Tri3d2Connectivity(
            [
                19,
                31,
                18,
            ],
        ),
        Tri3d2Connectivity(
            [
                21,
                32,
                20,
            ],
        ),
        Tri3d2Connectivity(
            [
                40,
                42,
                39,
            ],
        ),
        Tri3d2Connectivity(
            [
                42,
                43,
                39,
            ],
        ),
        Tri3d2Connectivity(
            [
                51,
                65,
                57,
            ],
        ),
        Tri3d2Connectivity(
            [
                14,
                28,
                13,
            ],
        ),
        Tri3d2Connectivity(
            [
                26,
                29,
                25,
            ],
        ),
        Tri3d2Connectivity(
            [
                13,
                55,
                12,
            ],
        ),
        Tri3d2Connectivity(
            [
                50,
                61,
                46,
            ],
        ),
        Tri3d2Connectivity(
            [
                27,
                56,
                26,
            ],
        ),
        Tri3d2Connectivity(
            [
                25,
                58,
                24,
            ],
        ),
        Tri3d2Connectivity(
            [
                15,
                59,
                14,
            ],
        ),
        Tri3d2Connectivity(
            [
                28,
                55,
                13,
            ],
        ),
        Tri3d2Connectivity(
            [
                46,
                61,
                47,
            ],
        ),
        Tri3d2Connectivity(
            [
                11,
                60,
                10,
            ],
        ),
        Tri3d2Connectivity(
            [
                26,
                56,
                29,
            ],
        ),
        Tri3d2Connectivity(
            [
                14,
                59,
                28,
            ],
        ),
        Tri3d2Connectivity(
            [
                29,
                58,
                25,
            ],
        ),
        Tri3d2Connectivity(
            [
                56,
                57,
                29,
            ],
        ),
        Tri3d2Connectivity(
            [
                52,
                57,
                56,
            ],
        ),
        Tri3d2Connectivity(
            [
                4,
                64,
                0,
            ],
        ),
        Tri3d2Connectivity(
            [
                16,
                62,
                2,
            ],
        ),
        Tri3d2Connectivity(
            [
                3,
                63,
                23,
            ],
        ),
        Tri3d2Connectivity(
            [
                47,
                69,
                35,
            ],
        ),
        Tri3d2Connectivity(
            [
                0,
                64,
                27,
            ],
        ),
        Tri3d2Connectivity(
            [
                2,
                62,
                15,
            ],
        ),
        Tri3d2Connectivity(
            [
                24,
                63,
                3,
            ],
        ),
        Tri3d2Connectivity(
            [
                34,
                66,
                41,
            ],
        ),
        Tri3d2Connectivity(
            [
                57,
                65,
                29,
            ],
        ),
        Tri3d2Connectivity(
            [
                36,
                65,
                51,
            ],
        ),
        Tri3d2Connectivity(
            [
                55,
                67,
                12,
            ],
        ),
        Tri3d2Connectivity(
            [
                1,
                67,
                11,
            ],
        ),
        Tri3d2Connectivity(
            [
                35,
                62,
                16,
            ],
        ),
        Tri3d2Connectivity(
            [
                23,
                63,
                36,
            ],
        ),
        Tri3d2Connectivity(
            [
                52,
                64,
                4,
            ],
        ),
        Tri3d2Connectivity(
            [
                58,
                65,
                36,
            ],
        ),
        Tri3d2Connectivity(
            [
                42,
                68,
                30,
            ],
        ),
        Tri3d2Connectivity(
            [
                60,
                67,
                55,
            ],
        ),
        Tri3d2Connectivity(
            [
                30,
                68,
                53,
            ],
        ),
        Tri3d2Connectivity(
            [
                53,
                66,
                34,
            ],
        ),
        Tri3d2Connectivity(
            [
                41,
                66,
                40,
            ],
        ),
        Tri3d2Connectivity(
            [
                53,
                68,
                66,
            ],
        ),
        Tri3d2Connectivity(
            [
                56,
                64,
                52,
            ],
        ),
        Tri3d2Connectivity(
            [
                66,
                68,
                40,
            ],
        ),
        Tri3d2Connectivity(
            [
                36,
                63,
                58,
            ],
        ),
        Tri3d2Connectivity(
            [
                59,
                62,
                35,
            ],
        ),
        Tri3d2Connectivity(
            [
                40,
                68,
                42,
            ],
        ),
        Tri3d2Connectivity(
            [
                50,
                60,
                55,
            ],
        ),
        Tri3d2Connectivity(
            [
                59,
                69,
                28,
            ],
        ),
        Tri3d2Connectivity(
            [
                12,
                67,
                1,
            ],
        ),
        Tri3d2Connectivity(
            [
                35,
                69,
                59,
            ],
        ),
        Tri3d2Connectivity(
            [
                27,
                64,
                56,
            ],
        ),
        Tri3d2Connectivity(
            [
                15,
                62,
                59,
            ],
        ),
        Tri3d2Connectivity(
            [
                58,
                63,
                24,
            ],
        ),
        Tri3d2Connectivity(
            [
                29,
                65,
                58,
            ],
        ),
        Tri3d2Connectivity(
            [
                28,
                69,
                61,
            ],
        ),
        Tri3d2Connectivity(
            [
                11,
                67,
                60,
            ],
        ),
        Tri3d2Connectivity(
            [
                61,
                69,
                47,
            ],
        ),
    ],
}
